Fix LibraryVersions.txt

fmahebert opened the pull request:
## Proposed changes

- Instead of appending each library's version and THEN writing the file anew (erasing all the contents), now write the file at the beginning of the build and then append each library's version
- Rename LibraryVersions.txt to BuildInfoAndVersions.txt to better reflect the content.

This closes #2645 with option (2) discussed in that issue.


### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

fmahebert committed:
Clear LibraryVersions.txt file at start of CMake

Previously, the version file was cleared at end of CMake, which would
erase any information about library versions written into the file
during the CMake configuration. By clearing the file at the start of
CMake, we keep all information from the configuration.

To streamline how things are written into the file, we use
PrintUsefulCMakeInfo to append build info to LibraryVersions. Since the
info being written to the LibraryVersions.txt file and the info being
printed to the command-line were nearly identical, just do both from the
same file. This is done at the end of CMake, and ensures the build info
being written into LibraryVersions.txt is up-to-date.

fmahebert committed:
Make stylistic changes to CMake outputs

nilsvu commented on cmake/PrintUsefulCMakeInfo.cmake:
```diff
@@ -1,14 +1,40 @@
 
+# First append useful info to BuildInfoAndLibraryVersions.txt
+file(APPEND
+  "${CMAKE_BINARY_DIR}/BuildInfoAndLibraryVersions.txt"
```
Just `BuildInfo.txt`? That's easier to tell people when they are supposed to send you the file :) and version info is also build info.

nilsvu commented on cmake/SetupBuildInfoAndVersionFile.cmake:
```diff
@@ -0,0 +1,14 @@
+  "SpECTRE Version: ${SPECTRE_VERSION}\n"
+  "Git description: ${GIT_DESCRIPTION}\n"
+  "Git branch: ${GIT_BRANCH}\n"
+  "Git hash: ${GIT_HASH}\n"
```
If you set up the file even earlier you can append the spectre version and git info in `SpectreLoadMetadata.cmake`. Up to you.

nilsvu commented on cmake/SetupCatch.cmake:
```diff
@@ -4,10 +4,10 @@
-message(STATUS "Catch include: ${CATCH_INCLUDE_DIR}")
-message(STATUS "Catch version: ${CATCH_VERSION}")
+message(STATUS "Catch incl: ${CATCH_INCLUDE_DIR}")
+message(STATUS "Catch vers: ${CATCH_VERSION}")
```
Why abbreviate? Better write "version" I think. (Also for Boost)

fmahebert replied:
For consistency with all our other SetupBlah cmake output 🙂... I would be just as happy to make everything output the unabbreviated text instead (vers/incl/libs -> version/include/library), but that's a bigger diff

nilsvu replied:
I think some of your diff in this PR uses `version`, so I suggest to just go with that. No need to refactor all occurrences in this PR.

fmahebert replied:
🤷 okay, reverting this change

nilsvu reviewed (changes requested):
Thanks for cleaning this up :)

nilsvu commented on cmake/AddHostToBuildInfoFile.cmake:
```diff
@@ -0,0 +1,10 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+file(APPEND
```
Perhaps name this file SetSiteName or something like that and include the `site_name` command, or the other way round: inline the `file` command in the root CMakeLists.txt and delete this file?

nilsvu commented on cmake/ClearBuildInfoFile.cmake:
```diff
@@ -0,0 +1,5 @@
+# See LICENSE.txt for details.
+
+# Create/clear file so contents can then be appended into it.
+file(WRITE "${CMAKE_BINARY_DIR}/BuildInfo.txt" "")
```
Just inline this command in the root CMakeLists.txt and delete this file? It's just one line.

nilsvu reviewed (dismissed):
LGTM, please squash. Here are some more suggestions (completely optional).

Would you mind updating the file name `LibraryVersions` > `BuildInfo` in `.github/ISSUE_TEMPLATE.md` as well?

fmahebert committed:
Rename LibraryVersions to better indicate contents

Rename happens in:
- CMake, where data is written to file
- documentation, where file is discussed
- Formaline, where file is used

fmahebert commented:
@nilsleiffischer Yes, good suggestions. On my branch (not yet pushed) I've also updated a lot of formaline-related code that used `LibraryVersions.txt` so it now uses `BuildInfo.txt`.

One question though: after renaming the CMake log file to `BuildInfo.txt`, there's a bit of a cognitive conflict with the unrelated `SetupBuildInfo.cmake`. We could rename the log file (again) BuildInfo -> CMakeInfo or BuildLog or similar... or alternatively rename SetupBuildInfo -> SetupInfoFromBuild (or better, I just can't think of a better name at the moment). Thoughts?

nilsvu commented:
`SetupInfoFromBuild` sound ok to me. Or `SetupInformer`, referencing the library name?

fmahebert committed:
Rename SetupBuildInfo.cmake to SetupInformer.cmake

fmahebert commented:
Squashed in the latest changes. I like your suggestion of `SetupInformer.cmake` so I went with that.

I've reorganized the commits since the first time you looked... after the various changes I felt this new organization was cleaner.

nilsvu commented on CMakeLists.txt:
```diff
@@ -32,10 +35,8 @@ endif ()
-
-include(SpectreInitializeVariables)
 include(SpectreGetGitHash)
+include(SetSiteName)
```
This should probably be named `SpectreSetSiteName`

nilsvu reviewed (changes requested):
LGTM, just this one thing.

Also, could you double-check that formaline stores the file content in the H5 file correctly? We test a formaline build on CI, but don't check the content of the build info. That would be a great addition to the CI job of course (invoke `--dump-build-info` and check the output is the same as the original BuildInfo.txt), but I'm not asking you do that in this PR.

fmahebert commented:
Renamed to `SpectreSetSiteName`, squashed directly.

fmahebert commented:
Test failure is a clang-tidy complaint from Formaline code that is unrelated to this PR... (I just renamed something in the file).

nilsvu commented on docs/Tutorials/Visualization.md:
```diff
@@ -101,8 +101,8 @@ cause the program to exit after the information has been dumped. You can use the
 `LD_LIBRARY_PATH`, `LIBRARY_PATH`, and `CMAKE_PREFIX_PATH` at link time. The
-flag `--dump-library-versions` will print to screen the contents of the
-`LibraryVersions.txt` file generated by CMake. The flag `--dump-environment`
+flag `--dump-build-info` will print to screen the contents of the
```
There's one more `--dump-library-version` in line 99 above

fmahebert replied:
Oops, that one has a typo (no final 's') and so I missed it in the find-replace... thanks for noticing!

nilsvu reviewed (changes requested):
squash directly

fmahebert commented:
Squashed in the latest rename fix. Thanks for the detailed reviews @nilsleiffischer!
